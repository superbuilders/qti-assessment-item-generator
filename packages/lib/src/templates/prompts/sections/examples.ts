import { existsSync, readFileSync } from "node:fs"
import * as path from "node:path"
import { resolveLibPath } from "@/internal/paths"

const POSITIVE_EXAMPLE_CACHE = new Map<string, string>()
const NEGATIVE_EXAMPLE_CACHE = new Map<string, string>()

const POSITIVE_ROOT = resolveLibPath("templates/prompts/examples/positive")
const POSITIVE_TEMPLATES_DIR = path.join(POSITIVE_ROOT, "templates")
const POSITIVE_NOTES_DIR = path.join(POSITIVE_ROOT, "notes")

const NEGATIVE_ROOT = resolveLibPath("templates/prompts/examples/negative")
const NEGATIVE_TEMPLATES_DIR = path.join(NEGATIVE_ROOT, "templates")
const NEGATIVE_NOTES_DIR = path.join(NEGATIVE_ROOT, "notes")

export function createPositiveExampleSection(names: readonly string[]): string {
	if (names.length === 0) return ""
	const rendered = names
		.map((name) => loadPositiveExample(name))
		.filter((block) => block.length > 0)
		.join("\n\n")
	if (rendered.length === 0) return ""
	return `### POSITIVE_EXAMPLES
<examples type="positive">
${rendered}
</examples>`
}

export function createNegativeExampleSection(names: readonly string[]): string {
	if (names.length === 0) return ""
	const rendered = names
		.map((name) => loadNegativeExample(name))
		.filter((block) => block.length > 0)
		.join("\n\n")
	if (rendered.length === 0) return ""
	return `### NEGATIVE_EXAMPLES
<examples type="negative">
${rendered}
</examples>`
}

function loadPositiveExample(name: string): string {
	const cached = POSITIVE_EXAMPLE_CACHE.get(name)
	if (cached) return cached

	const codePath = path.join(POSITIVE_TEMPLATES_DIR, `${name}.ts`)
	if (!existsSync(codePath)) {
		return ""
	}

	const notesPath = path.join(POSITIVE_NOTES_DIR, `${name}.md`)
	const code = readFileSync(codePath, "utf-8")
	const notes = existsSync(notesPath)
		? readFileSync(notesPath, "utf-8").trim()
		: ""

	const rendered = `<example kind="positive" name="${name}" source="${codePath}">
<code>
${code}
</code>
${notes ? `<notes>\n${notes}\n</notes>` : "<notes />"}
</example>`
	POSITIVE_EXAMPLE_CACHE.set(name, rendered)
	return rendered
}

function loadNegativeExample(name: string): string {
	const existing = NEGATIVE_EXAMPLE_CACHE.get(name)
	if (existing) return existing

	const codePath = path.join(NEGATIVE_TEMPLATES_DIR, `${name}.ts`)
	if (!existsSync(codePath)) {
		return ""
	}
	const notesPath = path.join(NEGATIVE_NOTES_DIR, `${name}.md`)
	const code = readFileSync(codePath, "utf-8")
	const notes = existsSync(notesPath)
		? readFileSync(notesPath, "utf-8").trim()
		: ""
	const rendered = `<example kind="negative" name="${name}" source="${codePath}">
<code>
${code}
</code>
${notes ? `<notes>\n${notes}\n</notes>` : "<notes />"}
</example>`
	NEGATIVE_EXAMPLE_CACHE.set(name, rendered)
	return rendered
}
