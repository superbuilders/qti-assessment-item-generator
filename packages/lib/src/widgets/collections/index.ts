import { allWidgetsCollection } from "@/widgets/collections/all"
import { fourthGradeMathCollection } from "@/widgets/collections/fourth-grade-math"
import { mathCoreCollection } from "@/widgets/collections/math-core"
import { scienceCollection } from "@/widgets/collections/science"
import { simpleVisualCollection } from "@/widgets/collections/simple-visual"
import { teksMath4Collection } from "@/widgets/collections/teks-math-4"

export const widgetCollections: {
	readonly all: typeof allWidgetsCollection
	readonly science: typeof scienceCollection
	readonly "simple-visual": typeof simpleVisualCollection
	readonly "fourth-grade-math": typeof fourthGradeMathCollection
	readonly "teks-math-4": typeof teksMath4Collection
	readonly "math-core": typeof mathCoreCollection
} = {
	all: allWidgetsCollection,
	science: scienceCollection,
	"simple-visual": simpleVisualCollection,
	"fourth-grade-math": fourthGradeMathCollection,
	"teks-math-4": teksMath4Collection,
	"math-core": mathCoreCollection
}

export type { WidgetCollection } from "@/widgets/collections/types"
