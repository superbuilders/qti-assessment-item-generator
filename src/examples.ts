import type { AssessmentItemInput } from "@/core/item"
import type { allWidgetsCollection } from "@/widgets/collections/all"
import type { WidgetTypeTupleFrom } from "@/widgets/collections/types"

export const allExamples: AssessmentItemInput<WidgetTypeTupleFrom<typeof allWidgetsCollection>>[] = []
