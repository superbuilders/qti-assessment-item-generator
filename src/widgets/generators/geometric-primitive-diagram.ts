import { z } from "zod"
import { CanvasImpl } from "../../utils/canvas-impl"
import { PADDING } from "../../utils/constants"
import { Path2D } from "../../utils/path-builder"
import { theme } from "../../utils/theme"
import type { WidgetGenerator } from "../types"

// Defines a point with an optional label. Coordinates are calculated automatically.
const createPointDefinitionSchema = () =>
	z
		.object({
			label: z
				.string()
				.nullable()
				.describe("The text label to display next to the point (e.g., 'P', 'Q'). Null for no label.")
		})
		.strict()
type Point = { x: number; y: number; label: string | null }

/**
 * Creates a diagram to render fundamental geometric primitives. This widget can display
 * a single point, a line segment between two points, a ray starting from a point, a line
 * passing through two points, or a curved arc. It is designed for basic geometric
 * illustrations where the focus is on the object and its labeled points rather than a
 * full coordinate plane. The layout is determined by high-level properties like rotation
 * and length, not explicit coordinates.
 */
export const GeometricPrimitiveDiagramPropsSchema = z
	.object({
		type: z.literal("geometricPrimitiveDiagram"),
		width: z.number().positive().describe("The total width of the SVG canvas in pixels."),
		height: z.number().positive().describe("The total height of the SVG canvas in pixels."),
		primitive: z
			.discriminatedUnion("type", [
				z
					.object({
						type: z.literal("point").describe("A single, isolated point, always centered in the diagram."),
						pointOne: createPointDefinitionSchema().describe("The point to be displayed.")
					})
					.strict(),
				z
					.object({
						type: z.literal("segment").describe("A line segment with two defined endpoints."),
						pointOne: createPointDefinitionSchema().describe("The starting point of the segment."),
						pointTwo: createPointDefinitionSchema().describe("The ending point of the segment."),
						rotation: z
							.number()
							.default(0)
							.describe("The rotation of the segment in degrees. 0 is horizontal, 90 is vertical."),
						length: z
							.number()
							.positive()
							.nullable()
							.describe("The length of the segment in pixels. If not provided, it defaults to 60% of the canvas width.")
					})
					.strict(),
				z
					.object({
						type: z.literal("ray").describe("A ray starting at one point and extending through another."),
						pointOne: createPointDefinitionSchema().describe("The starting point (endpoint) of the ray."),
						pointTwo: createPointDefinitionSchema().describe("A second point that defines the direction of the ray."),
						rotation: z
							.number()
							.default(0)
							.describe("The rotation of the ray in degrees. 0 is horizontal, 90 is vertical."),
						length: z
							.number()
							.positive()
							.nullable()
							.describe(
								"The distance in pixels between the start point and the direction point. Defaults to 60% of canvas width."
							)
					})
					.strict(),
				z
					.object({
						type: z
							.literal("line")
							.describe("An infinite line passing through two points, with arrowheads on both ends."),
						pointOne: createPointDefinitionSchema().describe("A first point that the line passes through."),
						pointTwo: createPointDefinitionSchema().describe("A second point that the line passes through."),
						rotation: z.number().default(0).describe("The rotation of the line in degrees."),
						length: z
							.number()
							.positive()
							.nullable()
							.describe(
								"The distance in pixels between the two defined points on the line. Defaults to 60% of canvas width."
							)
					})
					.strict(),
				z
					.object({
						type: z.literal("arc").describe("A curved arc segment."),
						pointOne: createPointDefinitionSchema().describe("The starting point of the arc."),
						pointTwo: createPointDefinitionSchema().describe("The ending point of the arc."),
						rotation: z.number().default(0).describe("The rotation of the baseline of the arc in degrees."),
						length: z
							.number()
							.positive()
							.nullable()
							.describe("The distance in pixels between the arc's endpoints. Defaults to 60% of canvas width."),
						bulge: z
							.number()
							.default(0.5)
							.describe(
								"Controls the curvature. 0 is a straight line, positive values bow 'up', negative values bow 'down' relative to the rotation."
							)
					})
					.strict()
			])
			.describe("The geometric primitive to be rendered in the diagram.")
	})
	.strict()

export type GeometricPrimitiveDiagramProps = z.infer<typeof GeometricPrimitiveDiagramPropsSchema>

/**
 * Helper to draw a point marker and its label.
 * @param canvas The canvas implementation to draw on.
 * @param point The point's data including calculated coordinates.
 * @param awayFrom An optional second point to position the label away from.
 */
const drawPointAndLabel = (canvas: CanvasImpl, point: Point, awayFrom?: Point) => {
	canvas.drawCircle(point.x, point.y, theme.geometry.pointRadius.base, {
		fill: theme.colors.black
	})

	if (point.label) {
		let labelX = point.x + 10
		let labelY = point.y - 10

		if (awayFrom) {
			const dx = point.x - awayFrom.x
			const dy = point.y - awayFrom.y
			const dist = Math.hypot(dx, dy)
			if (dist > 0) {
				const offsetX = (dx / dist) * 18
				const offsetY = (dy / dist) * 18
				labelX = point.x + offsetX
				labelY = point.y + offsetY
			}
		}

		canvas.drawText({
			x: labelX,
			y: labelY,
			text: point.label,
			fill: theme.colors.text,
			fontPx: theme.font.size.large,
			fontWeight: theme.font.weight.bold,
			anchor: "middle",
			dominantBaseline: "middle"
		})
	}
}

/**
 * Extends a line defined by two points to the boundaries of a given box.
 */
const extendLineToBox = (
	p1: Point,
	p2: Point,
	box: { x: number; y: number; width: number; height: number }
): { start: Point; end: Point } | null => {
	const dx = p2.x - p1.x
	const dy = p2.y - p1.y

	if (dx === 0 && dy === 0) return null

	const intersections: Point[] = []
	const checkIntersection = (x1: number, y1: number, x2: number, y2: number) => {
		const sx = x2 - x1
		const sy = y2 - y1
		const den = dx * sy - dy * sx
		if (den === 0) return
		const rx = x1 - p1.x
		const ry = y1 - p1.y
		const t = (rx * sy - ry * sx) / den // parameter along infinite line through p1->p2
		const s = (rx * dy - ry * dx) / den // parameter along segment [0,1]
		if (s >= 0 && s <= 1) {
			intersections.push({ x: p1.x + t * dx, y: p1.y + t * dy, label: null })
		}
	}

	checkIntersection(box.x, box.y, box.x + box.width, box.y) // Top
	checkIntersection(box.x, box.y, box.x, box.y + box.height) // Left
	checkIntersection(box.x + box.width, box.y, box.x + box.width, box.y + box.height) // Right
	checkIntersection(box.x, box.y + box.height, box.x + box.width, box.y + box.height) // Bottom

	if (intersections.length < 2) return null

	// Sort intersections by distance from p1 to ensure consistent start/end
	intersections.sort((a, b) => {
		const distA = Math.hypot(a.x - p1.x, a.y - p1.y)
		const distB = Math.hypot(b.x - p1.x, b.y - p1.y)
		return distA - distB
	})

	const start: Point = intersections[0]
	const end: Point = intersections[intersections.length - 1]
	return { start, end }
}

/**
 * For rays: extend from p1 toward p2 until the box boundary, ensuring directionality matches p2.
 */
const extendRayToBoxEnd = (
	p1: Point,
	p2: Point,
	box: { x: number; y: number; width: number; height: number }
): Point | null => {
	const line = extendLineToBox(p1, p2, box)
	if (!line) return null
	const dirX = p2.x - p1.x
	const dirY = p2.y - p1.y
	const toStartX = line.start.x - p1.x
	const toStartY = line.start.y - p1.y
	const toEndX = line.end.x - p1.x
	const toEndY = line.end.y - p1.y
	const dotStart = dirX * toStartX + dirY * toStartY
	const dotEnd = dirX * toEndX + dirY * toEndY
	// Choose the intersection that lies in the forward direction from p1 toward p2 (positive dot)
	if (dotStart >= 0 && dotEnd >= 0) {
		// Both forward; choose the farther one (larger projection)
		return dotEnd >= dotStart ? line.end : line.start
	}
	if (dotStart >= 0) return line.start
	if (dotEnd >= 0) return line.end
	// If neither is forward (degenerate), fall back to the farther endpoint
	const distStart = Math.hypot(toStartX, toStartY)
	const distEnd = Math.hypot(toEndX, toEndY)
	return distEnd >= distStart ? line.end : line.start
}

/**
 * Generates an SVG diagram for a single geometric primitive.
 */
export const generateGeometricPrimitiveDiagram: WidgetGenerator<typeof GeometricPrimitiveDiagramPropsSchema> = async (
	props
) => {
	const { width, height, primitive } = props

	const canvas = new CanvasImpl({
		chartArea: { left: 0, top: 0, width, height },
		fontPxDefault: 12,
		lineHeightDefault: 1.2
	})

	const BBOX = { x: PADDING, y: PADDING, width: width - 2 * PADDING, height: height - 2 * PADDING }
	const ARROW_ID = "primitive-arrow"
	const degToRad = (deg: number) => (deg * Math.PI) / 180

	canvas.addDef(
		`<marker id="${ARROW_ID}" viewBox="0 0 10 10" refX="9" refY="5" markerWidth="6" markerHeight="6" orient="auto-start-reverse"><path d="M 0 0 L 10 5 L 0 10 z" fill="${theme.colors.black}"/></marker>`
	)

	const cx = width / 2
	const cy = height / 2

	switch (primitive.type) {
		case "point": {
			const p1: Point = { x: cx, y: cy, label: primitive.pointOne.label }
			drawPointAndLabel(canvas, p1)
			break
		}
		case "segment":
		case "ray":
		case "line":
		case "arc": {
			const length = primitive.length ?? Math.min(width, height) * 0.6
			const rotationRad = degToRad(primitive.rotation)
			const halfLength = length / 2

			const p1: Point = {
				x: cx - halfLength * Math.cos(rotationRad),
				y: cy - halfLength * Math.sin(rotationRad),
				label: primitive.pointOne.label
			}
			const p2: Point = {
				x: cx + halfLength * Math.cos(rotationRad),
				y: cy + halfLength * Math.sin(rotationRad),
				label: primitive.pointTwo.label
			}

			if (primitive.type === "segment") {
				canvas.drawLine(p1.x, p1.y, p2.x, p2.y, {
					stroke: theme.colors.black,
					strokeWidth: theme.stroke.width.thick
				})
			} else if (primitive.type === "ray") {
				const rayEnd = extendRayToBoxEnd(p1, p2, BBOX)
				if (rayEnd) {
					canvas.drawLine(p1.x, p1.y, rayEnd.x, rayEnd.y, {
						stroke: theme.colors.black,
						strokeWidth: theme.stroke.width.thick,
						markerEnd: `url(#${ARROW_ID})`
					})
				}
			} else if (primitive.type === "line") {
				const extended = extendLineToBox(p1, p2, BBOX)
				if (extended) {
					canvas.drawLine(extended.start.x, extended.start.y, extended.end.x, extended.end.y, {
						stroke: theme.colors.black,
						strokeWidth: theme.stroke.width.thick,
						markerStart: `url(#${ARROW_ID})`,
						markerEnd: `url(#${ARROW_ID})`
					})
				}
			} else if (primitive.type === "arc") {
				const midX = (p1.x + p2.x) / 2
				const midY = (p1.y + p2.y) / 2
				// Perpendicular vector for bulge
				const perpX = -(p2.y - p1.y)
				const perpY = p2.x - p1.x
				const bulgeAmount = length * primitive.bulge * 0.5

				const controlX = midX + (perpX / length) * bulgeAmount
				const controlY = midY + (perpY / length) * bulgeAmount

				const path = new Path2D().moveTo(p1.x, p1.y).quadraticCurveTo(controlX, controlY, p2.x, p2.y)

				canvas.drawPath(path, {
					stroke: theme.colors.black,
					strokeWidth: theme.stroke.width.thick,
					fill: "none"
				})
			}

			// Draw points and labels on top for all linear/arc types
			drawPointAndLabel(canvas, p1, p2)
			drawPointAndLabel(canvas, p2, p1)
			break
		}
	}

	const { svgBody, vbMinX, vbMinY, width: finalWidth, height: finalHeight } = canvas.finalize(PADDING)

	return `<svg width="${finalWidth}" height="${finalHeight}" viewBox="${vbMinX} ${vbMinY} ${finalWidth} ${finalHeight}" xmlns="http://www.w3.org/2000/svg" font-family="${theme.font.family.sans}">${svgBody}</svg>`
}
