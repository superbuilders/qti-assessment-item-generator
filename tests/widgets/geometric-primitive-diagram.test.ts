import { describe, expect, test } from "bun:test"
import * as errors from "@superbuilders/errors"
import { geometricPrimitiveDiagramExamples } from "../../examples/geometric-primitive-diagram"
import type { Widget } from "../../src/widgets/registry"
import { generateWidget } from "../../src/widgets/widget-generator"

describe("Widget: geometric-primitive-diagram", () => {
    const examples = geometricPrimitiveDiagramExamples as unknown as Widget[]

    examples.forEach((props, index) => {
        test(`should produce consistent output for example #${index + 1}`, async () => {
            const result = await errors.try(generateWidget(props))
            if (result.error) throw result.error
            expect(result.data).toMatchSnapshot()
        })
    })
})


