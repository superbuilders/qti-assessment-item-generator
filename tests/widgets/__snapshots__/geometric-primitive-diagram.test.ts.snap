// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`Widget: geometric-primitive-diagram should produce consistent output for example #1 1`] = `"<svg width="340" height="340" viewBox="-20 -20 340 340" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="14"><defs><marker id="primitive-arrow" viewBox="0 0 10 10" refX="9" refY="5" markerWidth="6" markerHeight="6" orient="auto-start-reverse"><path d="M 0 0 L 10 5 L 0 10 z" fill="#000000"/></marker></defs><line x1="76.27631601399074" y1="201.62187927159414" x2="223.72368398600926" y2="98.37812072840586" stroke="#000000" stroke-width="2"/><circle cx="76.27631601399074" cy="201.62187927159414" r="4" fill="#000000"/><text x="61.531579216788884" y="211.94625512591296" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#333333">P</text><circle cx="223.72368398600926" cy="98.37812072840586" r="4" fill="#000000"/><text x="238.46842078321112" y="88.05374487408703" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#333333">S</text></svg>"`;

exports[`Widget: geometric-primitive-diagram should produce consistent output for example #2 1`] = `"<svg width="340" height="340" viewBox="-20 -20 340 340" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="14"><defs><marker id="primitive-arrow" viewBox="0 0 10 10" refX="9" refY="5" markerWidth="6" markerHeight="6" orient="auto-start-reverse"><path d="M 0 0 L 10 5 L 0 10 z" fill="#000000"/></marker></defs><line x1="90" y1="150" x2="210" y2="150" stroke="#000000" stroke-width="2"/><circle cx="90" cy="150" r="4" fill="#000000"/><text x="72" y="150" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#333333">P</text><circle cx="210" cy="150" r="4" fill="#000000"/><text x="228" y="150" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#333333">S</text></svg>"`;

exports[`Widget: geometric-primitive-diagram should produce consistent output for example #3 1`] = `"<svg width="340" height="340" viewBox="-20 -20 340 340" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="14"><defs><marker id="primitive-arrow" viewBox="0 0 10 10" refX="9" refY="5" markerWidth="6" markerHeight="6" orient="auto-start-reverse"><path d="M 0 0 L 10 5 L 0 10 z" fill="#000000"/></marker></defs><circle cx="150" cy="150" r="4" fill="#000000"/><text x="160" y="140" font-size="16" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#333333">Q</text></svg>"`;

exports[`Widget: geometric-primitive-diagram should produce consistent output for example #4 1`] = `"<svg width="340" height="340" viewBox="-20 -20 340 340" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="14"><defs><marker id="primitive-arrow" viewBox="0 0 10 10" refX="9" refY="5" markerWidth="6" markerHeight="6" orient="auto-start-reverse"><path d="M 0 0 L 10 5 L 0 10 z" fill="#000000"/></marker></defs><line x1="171.13091308703497" y1="195.3153893518325" x2="89.38000443985018" y2="20" stroke="#000000" stroke-width="2" marker-end="url(#primitive-arrow)"/><circle cx="171.13091308703497" cy="195.3153893518325" r="4" fill="#000000"/><circle cx="128.86908691296503" cy="104.6846106481675" r="4" fill="#000000"/></svg>"`;

exports[`Widget: geometric-primitive-diagram should produce consistent output for example #5 1`] = `"<svg width="340" height="340" viewBox="-20 -20 340 340" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="14"><defs><marker id="primitive-arrow" viewBox="0 0 10 10" refX="9" refY="5" markerWidth="6" markerHeight="6" orient="auto-start-reverse"><path d="M 0 0 L 10 5 L 0 10 z" fill="#000000"/></marker></defs><path d="M 103.01536896070458 167.10100716628344 Q 160.26060429977005 178.19077862357724 196.98463103929544 132.89899283371656" fill="none" stroke="#000000" stroke-width="2"/><circle cx="103.01536896070458" cy="167.10100716628344" r="4" fill="#000000"/><circle cx="196.98463103929544" cy="132.89899283371656" r="4" fill="#000000"/></svg>"`;

exports[`Widget: geometric-primitive-diagram should produce consistent output for example #6 1`] = `"<svg width="340" height="340" viewBox="-20 -20 340 340" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="14"><defs><marker id="primitive-arrow" viewBox="0 0 10 10" refX="9" refY="5" markerWidth="6" markerHeight="6" orient="auto-start-reverse"><path d="M 0 0 L 10 5 L 0 10 z" fill="#000000"/></marker></defs><line x1="225.0555349946513" y1="280" x2="74.94446500534868" y2="20.00000000000003" stroke="#000000" stroke-width="2" marker-start="url(#primitive-arrow)" marker-end="url(#primitive-arrow)"/><circle cx="175" cy="193.30127018922195" r="4" fill="#000000"/><circle cx="125" cy="106.69872981077806" r="4" fill="#000000"/></svg>"`;

exports[`Widget: geometric-primitive-diagram should produce consistent output for example #7 1`] = `"<svg width="340" height="340" viewBox="-20 -20 340 340" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="14"><defs><marker id="primitive-arrow" viewBox="0 0 10 10" refX="9" refY="5" markerWidth="6" markerHeight="6" orient="auto-start-reverse"><path d="M 0 0 L 10 5 L 0 10 z" fill="#000000"/></marker></defs><line x1="184.41458618106276" y1="199.14912265733952" x2="115.58541381893724" y2="100.85087734266048" stroke="#000000" stroke-width="2"/><circle cx="184.41458618106276" cy="199.14912265733952" r="4" fill="#000000"/><circle cx="115.58541381893724" cy="100.85087734266048" r="4" fill="#000000"/></svg>"`;

exports[`Widget: geometric-primitive-diagram should produce consistent output for example #8 1`] = `"<svg width="340" height="90" viewBox="-20 -20 340 90" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="14"><defs><marker id="primitive-arrow" viewBox="0 0 10 10" refX="9" refY="5" markerWidth="6" markerHeight="6" orient="auto-start-reverse"><path d="M 0 0 L 10 5 L 0 10 z" fill="#000000"/></marker></defs><line x1="20" y1="25" x2="280" y2="25" stroke="#000000" stroke-width="2" marker-start="url(#primitive-arrow)" marker-end="url(#primitive-arrow)"/><circle cx="37.5" cy="25" r="4" fill="#000000"/><circle cx="262.5" cy="25" r="4" fill="#000000"/></svg>"`;

exports[`Widget: geometric-primitive-diagram should produce consistent output for example #9 1`] = `"<svg width="340" height="90" viewBox="-20 -20 340 90" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="14"><defs><marker id="primitive-arrow" viewBox="0 0 10 10" refX="9" refY="5" markerWidth="6" markerHeight="6" orient="auto-start-reverse"><path d="M 0 0 L 10 5 L 0 10 z" fill="#000000"/></marker></defs><line x1="37.5" y1="25" x2="280" y2="25" stroke="#000000" stroke-width="2" marker-end="url(#primitive-arrow)"/><circle cx="37.5" cy="25" r="4" fill="#000000"/><circle cx="262.5" cy="25" r="4" fill="#000000"/></svg>"`;

exports[`Widget: geometric-primitive-diagram should produce consistent output for example #10 1`] = `"<svg width="340" height="90" viewBox="-20 -20 340 90" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="14"><defs><marker id="primitive-arrow" viewBox="0 0 10 10" refX="9" refY="5" markerWidth="6" markerHeight="6" orient="auto-start-reverse"><path d="M 0 0 L 10 5 L 0 10 z" fill="#000000"/></marker></defs><line x1="37.5" y1="25" x2="262.5" y2="25" stroke="#000000" stroke-width="2"/><circle cx="37.5" cy="25" r="4" fill="#000000"/><circle cx="262.5" cy="25" r="4" fill="#000000"/></svg>"`;

exports[`Widget: geometric-primitive-diagram should produce consistent output for example #11 1`] = `"<svg width="90" height="90" viewBox="-20 -20 90 90" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="14"><defs><marker id="primitive-arrow" viewBox="0 0 10 10" refX="9" refY="5" markerWidth="6" markerHeight="6" orient="auto-start-reverse"><path d="M 0 0 L 10 5 L 0 10 z" fill="#000000"/></marker></defs><circle cx="25" cy="25" r="4" fill="#000000"/></svg>"`;
